core/iwasm/libraries/lib-wasi-threads/stress_test/mutex_common.h

```diff
@@ -0,0 +1,261 @@
+        call_stat.tids[i] = 0;
+        call_stat.calls[i] = 0;
+    }
+}
+
+void
+add_to_stat(int tid)
+{
+    int tid_num = 0;
+    for (; tid_num < NUM_THREADS + 1 && call_stat.tids[tid_num] != 0;
+         ++tid_num) {
+        if (call_stat.tids[tid_num] == tid) {
+            call_stat.calls[tid_num]++;
+            return;
+        }
+    }
+
+    assert(tid_num < NUM_THREADS + 1);
+    call_stat.tids[tid_num] = tid;
+    call_stat.calls[tid_num] = 1;
+}
+
+// This function ensures that every of NUM_THREADS+1 threads number of
+// operations is in between [NUM_ITER/NUM_THREADS, NUM_ITER/(NUM_THREADS+2)].
+// For 100'000 operations it means that every thread should be called [7142,
+// 8333] times. Also, such a simple hash function might be not the ideal one so
+// the confidence interval is kept quite wide
+void
+check_stat()
+{
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        if (call_stat.calls[i] != 0) {
+            assert(call_stat.calls[i] < NUM_ITER / NUM_THREADS
+                   && "Call distribution by TID is out of range");
+            assert(call_stat.calls[i] > NUM_ITER / (NUM_THREADS + 2)
+                   && "Call distribution by TID is out of range");
+        }
+    }
+}
+
+void
+print_stat()
+{
+    fprintf(stderr, "Thread calls count by TID\n");
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        if (call_stat.tids[i] != 0) {
+            fprintf(stderr, "TID: %d; Calls: %d\n", call_stat.tids[i],
+                    call_stat.calls[i]);
+        }
+    }
+}
+
+// xor with a random value usually gives quite a uniform distribution
+uint32_t
+hash(uint32_t counter)
+{
+    return (counter << 22) ^ (0xA2BC22A); // Random value for hash func
+}
+
+// Locking and unlocking the same thread
+void
+same_thread_test(pthread_mutex_t *mutex)
+{
+    assert(pthread_mutex_lock(mutex) == 0
+           && "Main thread should be able to lock a mutex");
+    assert(pthread_mutex_unlock(mutex) == 0
+           && "Main thread should be able to unlock a mutex");
+}
+
+void *
+inc_until_limit(void *arg)
+{
+    MutexCounter mutex_counter = *(MutexCounter *)(arg);
+    int sleep_us = 0;
+    while (!pthread_mutex_lock(mutex_counter.mutex)
+           && *mutex_counter.counter < mutex_counter.limit) {
+        (*mutex_counter.counter)++;
+        add_to_stat((int)(pthread_self()));
+        if (mutex_counter.is_sleeping) {
+            sleep_us = hash(*mutex_counter.counter) % 1000;
+        }
+
+        assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+               && "Should be able to unlock a mutex");
+        if (mutex_counter.is_sleeping) {
+            usleep(sleep_us);
+        }
+    }
+
+    assert(*mutex_counter.counter == mutex_counter.limit);
+    assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+           && "Should be able to unlock the mutex after test execution");
+
+    return NULL;
+}
+
+void
+non_main_test(pthread_mutex_t *mutex, int num_iter)
```

Comment:
is this test conceptually the same as `same_thread_test` except that the code runs on a spawned thread?

Reply:
yeah
it can probably be just a flag (main / non-main)

Reply:
so both functions check if mutex can be locked / unlocked many times, it's just that one does it in a main thread, and the other does it in a non-main thread?

Reply:
yeah but i think it's a valid case to check. 

tell me if it's not needed

Reply:
well I decided to refactor it a bit anyway

Reply:
I refactored it to call the same func as in `same_thread_test` but from non-main thread